### PR TITLE
Issues#3256,#3255,#3252 Module bugfixes

### PR DESCRIPTION
**Pull request by suwc**
- Module import/export in if/else/while/for or function block should be syntax error (#3256)
 - Module should not allow 'return at top-level (#3255)
 - Module namespace should not have Symbol.iterator property (#3252)

### CODE REVIEW

**Comment by curtisman** [comment]
>     ListForListIterator* ModuleNamespace::EnsureSortedExportedNames()

No longer have any reference?  Remove?

---
Refers to: lib/Runtime/Language/ModuleNamespace.cpp:306 in 6ce8c61. [](commit_id = 6ce8c612f4ebb21393b61c148da73a93f8ef9f17, deletion_comment = False)

File: lib/Parser/Parse.cpp [hunk 1]
```diff
@@ -10106,7 +10115,7 @@ ParseNodePtr Parser::ParseStatement()
     {
         if (buildAST)
         {
-            if (nullptr == m_currentNodeFunc)
+            if (nullptr == m_currentNodeFunc || IsImportOrExportStatementValidHere())
```
Comment (curtisman) [inline lib/Parser/Parse.cpp]: This looks weird to me. 
Is this saying that `return` is an error, but 'try { return; } catch {}` is not?

**Comment by curtisman** [comment]
>         ParseNodePtr pnodeBody = ParseStatement<buildAST>();

How about this one?


---
Refers to: lib/Parser/Parse.cpp:9795 in 6ce8c61. [](commit_id = 6ce8c612f4ebb21393b61c148da73a93f8ef9f17, deletion_comment = False)

File: lib/Runtime/Language/ModuleNamespace.cpp [hunk 2]
```diff
@@ -338,28 +330,4 @@ namespace Js
         }
         return sortedExportedNames;
     }
-
-    Var ModuleNamespace::EntrySymbolIterator(RecyclableObject* function, CallInfo callInfo, ...)
```
Comment (boingoing) [inline lib/Runtime/Language/ModuleNamespace.cpp]: So without this, does property iteration still work just by doing an ordinary walk across the type handler?

Comment (suwc) [inline lib/Runtime/Language/ModuleNamespace.cpp]: Yes, property iteration still works.

Comment (boingoing) [inline lib/Runtime/Language/ModuleNamespace.cpp]: Does that affect the iteration order or does that continue to work because we already sorted them ?

Comment (suwc) [inline lib/Runtime/Language/ModuleNamespace.cpp]: This alone doesn't affect the iteration order. We have a separate issue to address the iteration order: [https://github.com/Microsoft/ChakraCore/issues/3246](https://github.com/Microsoft/ChakraCore/issues/3246)

---
In reply to: [126488250](https://github.com/Microsoft/ChakraCore/pull/3307#discussion_r126488250) [](ancestors = 126488250)

File: lib/Parser/Parse.cpp (hunk 1 above)
Comment (suwc) [inline lib/Parser/Parse.cpp]: Clarified 'try { return }' and others (except inside a function body) should be treated as top-level return. Fixed in 2nd iteration.

---
In reply to: [126453545](https://github.com/Microsoft/ChakraCore/pull/3307#discussion_r126453545) [](ancestors = 126453545)

**Comment by suwc** [comment]
Any other comments?

**Comment by boingoing** [comment]
Looks good to me. I wondered if `GetCurrentFunctionNode()` can return null but I see it used all over the parser with loose checking so it's probably fine.

**Comment by pleath** [comment]
GetCurrentFunctionNode returns the current enclosing function node, of which there's always one, even if it's the one that encloses the global statements.

**Pull request merged**